
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ #
#                                   INSTALLATION                                  # 
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ #
# From Bioconductor
source("http://www.bioconductor.org/biocLite.R")
biocLite("GRridge")

# From github
library(devtools)
install_github("markvdwiel/GRridge")

library(GRridge)
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ #



# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ #
#                                   DATA PREPARATION                              #
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ #
# classification case (response): breast and colorectal cancer (CRC)

# Load the primary data set 
data(dataWurdinger)

# Transform the data set to the square root scale
dataSqrtWurdinger <- sqrt(datWurdinger_BC)

#Standardize the transformed data
datStdWurdinger <- t(apply(dataSqrtWurdinger,1,
                           function(x){(x-mean(x))/sd(x)}))

# A list of gene names in the primary RNAseq data
genesWurdinger <- annotationWurdinger$geneSymbol
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ #


# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ #
#                     PARTITIONS CONTAINING OVERLAPPING GROUPS                    #
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ #

# 1. TRANSCRIPTION FACTOR BINDING SITE PATHWAY
# Gene sets contain genes that share a transcription factor binding site 
# defined in the TRANSFAC (version 7.4, http://www.gene-regulation.com/) database. 
# Each of these gene sets is annotated by a TRANSFAC record.
# see: https://github.com/markvdwiel/GRridgeCodata/tree/master/Transcription-factor-binding-site-pathway
# where the data is also available

# Exporting a GMT file
TFsym <- getGmt("c3.tft.v5.0.symbols.gmt")

# A partition based on the GSEA object (IMsym) is then created. 
# Some features may belong to more than one group. 
# The argument minlen=25 implies the minimum number of features in a gene set. 
# If remain=TRUE, gene sets with less than 25 members are grouped to the 
# "remainder" group. 
gseTF <- matchGeneSets(genesWurdinger,TFsym,minlen=25,remain=TRUE)

# Merge groups in a partition
# Pathway-based partition often contains a considerable number of gene sets. 
# There are 615 and 4871 groups in the transcription factor and immunological 
# based pathway, respectively (per July 4, 2016).  Overfitting may be an issue 
# in the GRridge predictive modeling on such a large number of groups. 
# As a solution, a data driven re-grouping based on hierarchical clustering analysis 
# can be applied. The GRridge package provides a function to merge groups in a 
# partition,i.e. mergeGroups. In this example  the initial gene sets will be
# re-grouped into 5 groups (maxGroups=5). 
gseTF_newGroups <- mergeGroups(highdimdata= datStdWurdinger, 
                               initGroups=gseTF, maxGroups=5)

# To extract indices of new groups
# The gseTF2 object is a list of the components of which contain the indices 
# of the features belonging to each group. This object is in the same format as 
# the output from the "CreatePartition" function. Hence, the result can be used
# directly as an input in the "grridge" function. 
parTranscriptFactor <- gseTF_newGroups$newGroups

# To observe members of the new groups
newClustMembers <- gseTF_newGroups$newGroupMembers



# 2. IMMUNOLOGIC SIGNATURE PATHWAY
# Gene sets that represent cell states and perturbations within the immune system.
# The signatures were generated by manual curation of published studies in human 
# and mouse immunology. For each study, pairwise comparisons of relevant classes 
# were made and genes ranked by mutual information. Gene sets correspond to top or
# bottom genes (FDR < 0.25 or maximum of 200 genes) for each comparison.
# see: https://github.com/markvdwiel/GRridgeCodata/tree/master/Immunologic-signature-pathway
# for more detail

# following the same procedure as the previous partition
IMsym <- getGmt("c7.all.v5.1.symbols.gmt")
gseIM <- matchGeneSets(genesWurdinger,IMsym,minlen=25,remain=TRUE)
gseIM_newGroups <- mergeGroups(highdimdata= datStdWurdinger, initGroups=gseIM, 
                               maxGroups=5)
parImmun  <- gseIM_newGroups$newGroups
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ #



# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ #
#            A PARTITION BASED ON A LIST OF PLATELETS EXPRESSED GENES             #
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ #
# The list is available at
# https://github.com/markvdwiel/GRridgeCodata/blob/master/Platelet-expressed-genes/PlateletExpressedGenesList.txt
plateletsExprGenes <- as.character(read.table("PlateletExpressedGenesList.txt"))

# Group genes in the primary data based on the list
# The genes are grouped into either "NormalGenes" or "Non-overlapGenes"
is <- intersect(plateletsExprGenes,genesWurdinger)
im <- match(is, genesWurdinger)
plateletsGenes <- replicate(length(genesWurdinger),"Non-overlapGenes")
plateletsGenes[im] <- "NormalGenes"
plateletsGenes <- as.factor(plateletsGenes)
parPlateletGenes <- CreatePartition(plateletsGenes)
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ #



# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ #
#                    A PARTITION BASED ON CHROMOSOMAL LOCATION                    #
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ #
# A list of chromosomal location based on biomaRt data bases.
ChromosomeWur0 <- as.vector(annotationWurdinger$chromosome)
ChromosomeWur <- ChromosomeWur0
idC <- which(ChromosomeWur0=="MT" | ChromosomeWur0=="notBiomart" |
               ChromosomeWur0=="Un")
ChromosomeWur[idC] <- "notMapped"
table(ChromosomeWur)
parChromosome <- CreatePartition(as.factor(ChromosomeWur))
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ #



# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ #
#                   CONCATENATE ALL PARTITIONS INTO ONE LIST                      #
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ #
partitionWurdinger <- list(immunPathway=parImmun, TFPathway= parTranscriptFactor, 
                           plateletsGenes=parPlateletGenes, chromosome=parChromosome)

# A list of monotone functions from the corresponding partitions,
monotoneWurdinger <- c(FALSE,FALSE,FALSE,FALSE)
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ #



# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ #
#                             PARTITION SELECTION PROCEDURE                       #
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ #
# It may take some time!
optPartitions <- PartitionsSelection(datStdWurdinger, respWurdinger,
                                     partitions=partitionsWurdinger,
                                     monotoneFunctions=monotoneWurdinger)
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ #



# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ #
#                                  GRridge MODEL                                  #
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ #
# As comparison, lasso model is also built (comparelasso=TRUE) and post-hoc 
# feature selection via L1 penalization method is performed (selectionEN=TRUE)
# by predetermined the number of selected markers (maxsel=10).

# To reduce the computational time, we may use the optimum lambda2
# (global lambda penalty) in the GRridge predictive modeling,
# optl=optPartitions$optl
# GRridge model by incorporating the selected partitions
partitionsWurdinger_update = partitionsWurdinger[optPartitions$ordPar]
monotoneWurdinger_update = monotoneWurdinger[optPartitions$ordPar]

grWurdinger <- grridge(datStdWurdinger,respWurdinger,
                       partitions=partitionsWurdinger_update,
                       monotone= monotoneWurdinger_update,innfold = 3,
                       comparelasso=TRUE,optl=optPartitions$optl,
                       selectionEN=TRUE,maxsel=10)

# Asses the performance of the GRridge model by performing 10-fold CV
grWurdingerCV <- grridgeCV(grWurdinger, datStdWurdinger, respWurdinger,
                           outerfold=10)
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ #



# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ #
#                                 MODEL ASSESSMENT                                #
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ #
cutoffs <- rev(seq(0,1,by=0.01))
rocridge <- roc(probs= grWurdingerCV[,2],
                true= grWurdingerCV[,1],cutoffs)
rocGRridge <- roc(probs= grWurdingerCV[,3],
                  true= grWurdingerCV[,1],cutoffs)
rocLasso <- roc(probs= grWurdingerCV[,4],
                true= grWurdingerCV[,1],cutoffs)
rocGRridgeEN <- roc(probs= grWurdingerCV[,5],
                    true= grWurdingerCV[,1],cutoffs)
plot(rocridge[1,],rocridge[2,],type="l",lty=2,ann=FALSE,col="grey")
points(rocGRridge[1,],rocGRridge[2,],type="l",lty=1,col="black")
points(rocLasso[1,],rocLasso[2,],type="l",lty=1,col="blue")
points(rocGRridgeEN[1,],rocGRridgeEN[2,],type="l",lty=1,col="green")
legend(0.6,0.35, legend=c("ridge","GRridge", "lasso","GRridge+varsel"),
       lty=c(1,1), lwd=c(1,1),col=c("grey","black","blue","green"))
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ #
